test/protected_resource_test.rb

```diff
@@ -112,4 +112,43 @@ def test_responds_to_name_when_bulkhead_or_circuit_breaker_disabled
     assert_equal :no_bulkhead, Semian[:no_bulkhead].name
     assert_equal :no_circuit_breaker, Semian[:no_circuit_breaker].name
   end
+
+  def test_gracefully_fails_when_unable_to_decrement_ticket_count
```

Comment:
I'm pretty satisfied with it, the only downside is that the timeout for the internal semaphore is 5s and not configurable, so it does take 5s.